File: tests/fixtures/slickrun.scala
// Regression fixture for the `agent/slickrun` slice: the shapes that made
// scala-rs-compiled slick fail at *run* time even though every class file
// loaded. One file, many cases, in the order they were found.
import scala.collection.mutable

// (1) A nested `def` whose `match` binds names: those binders are bound, not
// free. Counting them as captures grew the lifted method extra parameters and
// made the enclosing trait declare capture accessors it could not implement.
class Nd
class CS(val sql: String) extends Nd
case class PS(cases: Seq[(Int => Boolean, Nd)], default: Nd) extends Nd
trait Comp {
  class Ext(tree: Nd, param: Int) {
    def result: String = {
      def findSql(n: Nd): String = n match {
        case c: CS => c.sql
        case PS(cases, default) =>
          findSql(cases.find { case (f, n) => f(param) }.map(_._2).getOrElse(default))
        case _ => "?"
      }
      findSql(tree)
    }
  }
}

// (2) An existential's skolem erases to its upper bound, so a member read
// through `Box[_ <: E]` owes a `checkcast` to `E`'s bound.
class Box[T](val value: T) { def get2: T = value }
abstract class Base { def tag: String }
class Impl(val tag: String) extends Base
class Holder[E <: Base](e: E) {
  lazy val shaped: Box[_ <: E] = new Box[E](e)
  def baseRow: E = shaped.value
  def baseRow2: E = shaped.get2
}

// (3) A case class's companion is not the case class: it must not run the
// `$init$` of the traits the class mixes in.
trait Marked { val mark: String = "m" }
final case class Ap(a: Int) extends Marked
final case class Ap2(a: Int, b: Int)(val extra: String) extends Marked

// (4) `3.compare(4)` goes through `RichInt`, not through an unmaterialised
// `Ordered` view.
class Counter {
  def length: Int = 3
  def lengthCompare(n: Int): Int = length.compare(n)
}

// (5) An auxiliary constructor of an *inner* class takes the enclosing
// instance too, and `this(...)` passes it on.
class Outer {
  class Inner(val a: Int, val b: String) {
    def this(a: Int) = this(a, "d")
    def show = a + b + Outer.this.tag
  }
  def tag = "!"
  def mk = new Inner(7)
}

// (6) `FunctionN.apply` erases to `Object`: a *tuple* result needs the cast
// its own method descriptor promises.
class T1 { override def toString = "T" }
class U1 { override def toString = "U" }

// (7) A trait `val` overridden by a narrower one in a derived trait: the base
// trait's mixin setter still has to exist (as a no-op) and the getter needs a
// bridge.
trait Opts { def name: String }
class BaseOpts extends Opts { def name = "base" }
class SubOpts extends BaseOpts { override def name = "sub" }
trait HasOpts { val opts: BaseOpts = new BaseOpts }
trait HasSubOpts extends HasOpts { override val opts: SubOpts = new SubOpts }

// (8) A default argument on a *trait* method needs its `name$default$n`
// getter on the interface and on every implementing class.
trait Defaulted {
  def label(prefix: String, upper: Boolean = false): String =
    if (upper) prefix.toUpperCase else prefix
}
class Defaults extends Defaulted

// (9) A `private` constructor the companion calls is emitted public, like nsc
// does (a constructor cannot be renamed).
final class Wrapped private (val n: Int) {
  private def this() = this(0)
  def bump = new Wrapped()
}
object Wrapped { def make(n: Int) = new Wrapped(n) }

// (10) A `case class` pattern that names only the first parameter list reads
// the constructor fields; there is no companion `unapply` to call.
final case class Tbl(schema: Option[String], name: String)(val extra: Int)

// (11) `if (c) e` with no `else` and a non-`Unit` branch value: both paths
// have to leave the same stack height.
class OneArmed {
  var seen = 0
  def touch(n: Int): String = { seen += n; "t" + n }
  def run(c: Boolean): Any = { if (c) touch(1); seen }
}

object Main {
  def g(n: T1)(f: T1 => (T1, U1)): (T1, U1) = n match { case null => null; case x => f(x) }

  def main(args: Array[String]): Unit = {
    val comp = new Comp {}
    println(new comp.Ext(new CS("q"), 1).result)
    println(new comp.Ext(PS(Seq(((i: Int) => i > 0, new CS("hit"))), new CS("miss")), 1).result)

    val h = new Holder[Impl](new Impl("hi"))
    println(h.baseRow.tag + h.baseRow2.tag)

    println(Ap(1).mark + Ap.apply(3) + Ap2(1, 2)("x").extra)

    println(new Counter().lengthCompare(5))
    println(new Counter().lengthCompare(3))

    val o = new Outer
    println(o.mk.show + new o.Inner(3, "x").show)

    println(g(new T1)(t => (t, new U1)))

    val so: HasOpts = new HasSubOpts {}
    println(so.opts.name)

    val d = new Defaults
    println(d.label("ab") + d.label("ab", upper = true))

    println(Wrapped.make(4).n + Wrapped.make(4).bump.n)

    val t = Tbl(Some("s"), "t")(9)
    val shown = t match { case Tbl(s, n) => s.getOrElse("-") + n }
    println(shown + t.extra)

    val oa = new OneArmed
    println(oa.run(true).toString + oa.run(false).toString)

    val buf = mutable.ArrayBuffer(3, 1, 2)
    println(buf.sortInPlace().mkString(","))
  }
}

File: crates/cli/tests/slickrun.rs
//! E2E tests for the `agent/slickrun` slice (fixture prefix `slickrun`).
//!
//! This slice's question was not "does slick compile?" -- it already did, with
//! every one of its 4552 class files loading under `-Xverify:all` -- but "does
//! what came out actually *run*?". `tests/slick_run.sh` answers it by compiling
//! slick twice (scala-rs and real scalac), compiling a set of ordinary slick
//! client programs *once* with scalac, and running that one client binary
//! against each slick build. Every difference is therefore slick's class
//! files, i.e. scala-rs.
//!
//! The fixture collects the shapes that broke, each of which was a run-time
//! failure of code that had type-checked and verified in isolation:
//!
//!  1. a nested `def` whose `match` binds names -- the binders were counted as
//!     free variables, so the lifted method grew parameters for them and the
//!     enclosing trait declared capture accessors no class could implement
//!     (`throw new RuntimeException("cannot capture cons for trait
//!     JdbcProfile")`);
//!  2. `Box[_ <: E].value` -- an existential's skolem erases to its upper
//!     bound, and without the `checkcast` slick's `def baseTableRow: E =
//!     shaped.value` failed verification, killing every `TableQuery`;
//!  3. a case class's companion running the `$init$` of the traits the *class*
//!     mixes in (`IncompatibleClassChangeError` on `slick.ast.Apply$`);
//!  4. `length.compare(n)` reaching `Ordered` through a view that was never
//!     materialised, leaving a `checkcast scala/math/Ordered` on an `int`;
//!  5. an auxiliary constructor of an inner class missing its `$outer`
//!     parameter (`NoSuchMethodError` on the first `Table` subclass);
//!  6. a tuple result of `FunctionN.apply` left uncast;
//!  7. a trait `val` overridden by a narrower one in a derived trait -- the
//!     base trait's mixin setter and the getter bridge were both missing
//!     (`AbstractMethodError` initialising `H2Profile$`);
//!  8. a default argument on a *trait* method with no `name$default$n` getter
//!     anywhere;
//!  9. a `private` constructor the companion calls, emitted `ACC_PRIVATE`;
//! 10. a `case class` pattern naming only the first of several parameter lists,
//!     which took the extractor path to a companion `unapply` that is a symbol
//!     with no method behind it;
//! 11. `if (c) e` with no `else` and a non-`Unit` branch value, whose two paths
//!     met the join at different stack heights.
//!
//! Self-contained (own copies of the small helpers `e2e.rs` also has) per
//! `.agent-brief.md`: the shared test files belong to other in-flight agents.

use std::fs;
use std::path::PathBuf;
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};

fn bin() -> PathBuf {
    PathBuf::from(env!("CARGO_BIN_EXE_scala-rs"))
}

fn fixtures_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../tests/fixtures")
}

fn tmp_dir(tag: &str) -> PathBuf {
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    static SEQ: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(0);
    let seq = SEQ.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
    let p = std::env::temp_dir().join(format!(
        "scala-rs-slickrun-{tag}-{}-{nanos}-{seq}",
        std::process::id()
    ));
    fs::create_dir_all(&p).unwrap();
    p
}

fn expected_stdout(name: &str) -> String {
    fs::read_to_string(fixtures_dir().join("expected").join(format!("{name}.txt"))).unwrap()
}

fn java_available() -> bool {
    Command::new("java")
        .arg("-version")
        .output()
        .map(|o| o.status.success() || !o.stderr.is_empty() || !o.stdout.is_empty())
        .unwrap_or(false)
}

fn scala_library_jar() -> Option<PathBuf> {
    let cached = PathBuf::from("/tmp/scala-rs-lib/scala-library-2.13.16.jar");
    if cached.is_file() {
        return Some(cached);
    }
    let _ = fs::create_dir_all("/tmp/scala-rs-lib");
    let url = "https://repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.16/scala-library-2.13.16.jar";
    let status = Command::new("curl")
        .args(["-fsSL", "-o", cached.to_str().unwrap(), url])
        .status();
    if status.map(|s| s.success()).unwrap_or(false) && cached.is_file() {
        return Some(cached);
    }
    None
}

fn compile_fixture_with(name: &str, extra: &[&str]) -> PathBuf {
    let src = fixtures_dir().join(format!("{name}.scala"));
    let out = tmp_dir(name);
    let mut cmd = Command::new(bin());
    cmd.args([
        "compile",
        src.to_str().unwrap(),
        "-d",
        out.to_str().unwrap(),
    ]);
    cmd.args(extra);
    let status = cmd.status().expect("run scala-rs compile");
    assert!(status.success(), "compile {name} failed extra={extra:?}");
    out
}

/// Compile against the real jar and run under `-Xverify:all`. Every case in
/// the fixture is a *verification or linkage* failure when it regresses, so
/// the run is the test; the recorded expectation is real scalac 2.13.16's own
/// stdout for the same file.
fn dual_run_fixture(name: &str) {
    if !java_available() {
        return;
    }
    let Some(jar) = scala_library_jar() else {
        eprintln!("skip scala-library dual-run: jar not obtainable");
        return;
    };
    let jar_s = jar.to_str().unwrap();
    let out = compile_fixture_with(name, &["--scala-library", jar_s]);
    let cp = format!("{}:{}", out.display(), jar.display());
    let output = Command::new("java")
        .args(["-Xverify:all", "-cp", &cp, "Main"])
        .output()
        .expect("java");
    assert!(
        output.status.success(),
        "java -Xverify:all -cp out:scala-library failed for {name}: {}",
        String::from_utf8_lossy(&output.stderr)
    );
    assert_eq!(
        String::from_utf8_lossy(&output.stdout),
        expected_stdout(name),
        "stdout mismatch for library dual-run {name}"
    );
    let _ = fs::remove_dir_all(&out);
}

#[test]
fn fixtures_slickrun_dual_run() {
    dual_run_fixture("slickrun");
}

/// The four `{ case … }` literals in `PatLambdas` / `PatInTrait` are lowered to
/// hoisted `$anonfun$` statics, not to closure classes, and the run above shows
/// their pattern binders, the captured enclosing `this` and a captured `var`
/// all still line up in that static method's locals.
///
/// This is the shape `agent/slickrun` first saw fail — but the cause was a
/// merge artifact in `load_this`, not the lowering. Pinned here so the next
/// slice that widens the indy boundary (a `{ case … }` where a plain
/// `Function1` is expected) has coverage for it rather than a rumour.
#[test]
fn fixtures_slickrun_pattern_lambdas_are_hoisted() {
    if !java_available() {
        return;
    }
    let Some(jar) = scala_library_jar() else {
        return;
    };
    let out = compile_fixture_with("slickrun", &["--scala-library", jar.to_str().unwrap()]);
    let names: Vec<String> = fs::read_dir(&out)
        .unwrap()
        .filter_map(|e| e.ok())
        .map(|e| e.file_name().to_string_lossy().into_owned())
        .collect();
    let lambda_classes: Vec<&String> = names.iter().filter(|n| n.contains("anonfun")).collect();
    assert!(
        lambda_classes.is_empty(),
        "a lambda became a class file: {lambda_classes:?}"
    );
    // The only anonymous *classes* here are the two the source writes out --
    // `new Comp {}` and `new HasSubOpts {}`. If a `{ case … }` regresses to the
    // closure-class path this number moves, so move it deliberately.
    let anon: Vec<&String> = names.iter().filter(|n| n.contains("$$anon$")).collect();
    assert_eq!(
        anon.len(),
        2,
        "expected exactly two anonymous classes, got {anon:?}"
    );
    for cls in ["PatLambdas.class", "PatInTrait.class"] {
        let bytes = fs::read(out.join(cls)).unwrap_or_else(|e| panic!("{cls}: {e}"));
        let needle = b"$anonfun$";
        assert!(
            bytes.windows(needle.len()).any(|w| w == needle),
            "{cls} has no hoisted $anonfun$ method"
        );
    }
    let _ = fs::remove_dir_all(&out);
}
